shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs

```diff
@@ -36,97 +35,79 @@ public static string GetTypeDisplayName(object item, bool fullName = true)
+        private static void ProcessTypeName(StringBuilder builder, Type type, bool fullName)
         {
-            var totalArgs = args.Length;
-            if (totalArgs >= startIndex + numberOfArgsToAppend)
+            if (type.IsGenericType)
             {
-                sb.Append("<");
-                for (int i = startIndex; i < startIndex + numberOfArgsToAppend; i++)
-                {
-                    ProcessTypeName(args[i], sb, fullName);
-                    if (i + 1 < startIndex + numberOfArgsToAppend)
-                    {
-                        sb.Append(", ");
-                    }
-                }
-                sb.Append(">");
+                ProcessGenericType(builder, type, new ArraySegment<Type>(type.GetGenericArguments()), fullName);
+            }
+            else if (_builtInTypeNames.ContainsKey(type))
+            {
+                builder.Append(_builtInTypeNames[type]);
+            }
+            else
+            {
+                builder.Append(fullName ? type.FullName : type.Name);
             }
         }
 
-        private static void ProcessTypeName(Type t, StringBuilder sb, bool fullName)
+        private static void ProcessGenericType(StringBuilder builder, Type type, ArraySegment<Type> genericArguments, bool fullName)
         {
-            if (t.GetTypeInfo().IsGenericType)
+            var ownGenericArguments = genericArguments;
+            if (type.IsNested)
             {
-                ProcessNestedGenericTypes(t, sb, fullName);
-                return;
+                var offset = type.DeclaringType.GetGenericArguments().Length;
+                ownGenericArguments = new ArraySegment<Type>(genericArguments.Array, offset, genericArguments.Count - offset);
+            }
+
+            if (fullName)
+            {
+                if (type.IsNested)
+                {
+                    var declaringTypeGenericArguments = new ArraySegment<Type>(
+                        genericArguments.Array,
+                        0,
+                        genericArguments.Count - ownGenericArguments.Count);
+                    ProcessGenericType(builder, type.DeclaringType, declaringTypeGenericArguments, fullName);
+                    builder.Append('+');
+                }
+                else
+                {
+                    builder.Append(type.Namespace);
+                    builder.Append('.');
+                }
             }
-            if (_builtInTypeNames.ContainsKey(t))
+
+            var genericPartIndex = type.Name.IndexOf('`');
+            if (genericPartIndex > 0)
             {
-                sb.Append(_builtInTypeNames[t]);
+                builder.Append(type.Name, 0, genericPartIndex);
             }
             else
             {
-                sb.Append(fullName ? t.FullName : t.Name);
+                builder.Append(type.Name);
```

Comment:
We can put return here itself because if there are no generic parts there is nothing to print further. That will remove next if block.